### PR TITLE
SNOW-303944 Add utility function to python connector to retrieve JWT token from privatekey file and pwd

### PR DESCRIPTION
**Pull request by sfc-gh-mknister**
SNOW-303944
This change creates a static function **get_token_from_private_key()** that can be called by Snowsql to get a jwt token.

### CODE REVIEW

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/auth_keypair.py:47]: I'd rather make this an argument for the `__init__` function. Could we do that instead?
It could be something like:
```
def __init__(self, private_key, lifetime_in_seconds: int = 120):
...
self._lifetime = timedelta(seconds=lifetime_in_seconds)
```

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/auth_keypair.py:49]: Please use type hints for every function that you touch. [More info](https://docs.python.org/3/library/typing.html)
For this function it should be like:
```suggestion
    def change_token_lifetime(self, lifetime_in_seconds: int) -> None:
```

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/auth.py:476]: If this function doesn't use the class itself please don't make it a static function.
In Python not everything has to be in a class and if you wanted to use it from the outside it'd still be `snowflake.connector.auth.get_token_from_pricate_key`
Also add type hints please!

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/auth.py:477]: Type hints could help with this. If you declare `def get_token_from_private_key(user, account, privatekey_path, key_password: str):` then that says that `key_password` should not be None ever.

**Review by sfc-gh-mkeller — DISMISSED** [review]

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth_keypair.py]: Makes sense! Added

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth_keypair.py]: Removed this function entirely since now it's set in the constructor for KeyPairAuth

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py]: Removed- now it's just at the bottom of the class but not using the class object

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py]: I added type hints for user, account, and privatekey_path, but it's okay if key_password is None so I didn't add the type hint here. This could happen if the user has an unencrypted private key file and so doesn't set the environment variable SNOWSQL_PRIVATE_KEY_PASSPHRASE, which will then be returned as None.

Comment (sfc-gh-mkeller) [reply src/snowflake/connector/auth.py]: The right type hint for a nullable string is `Optional[str]`.

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/auth_keypair.py:38]: Add a docstring description for `lifetime_in_seconds` please

**Review by sfc-gh-mkeller — DISMISSED** [review]

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py]: Added!

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py:693]: The precommit hook called 'black' caused this to fail because I think comments inside the parameter list is not acceptable, but I added a comment describing the behavior above this line.

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-stakeda) [inline src/snowflake/connector/auth_keypair.py:38]:
```suggestion
    def __init__(self, private_key, lifetime_in_seconds: int = self.LIFETIME):
```

File: src/snowflake/connector/auth.py
Comment (sfc-gh-stakeda) [inline src/snowflake/connector/auth.py:694]: `authenticate` method should do the following:
```
if not '.global' in account:
    idx = account.find('.')
    if idx > 0:
        account = account[0:idx]
else:
    idx = account.find('-')
    if idx > 0:
        account = account[0:idx]
account = account.upper().replace('-', '_')
```

Comment (sfc-gh-stakeda) [reply src/snowflake/connector/auth.py:694]: Please add the test to cover all cases.

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/auth.py:693]: `black` will reformat your files, just run it again with the same commit message, or you can actually manually run `black` with `tox -e fix_lint` once you install `tox` with `python3 -m pip install tox`.

File: src/snowflake/connector/auth.py
Comment (sfc-gh-stakeda) [inline src/snowflake/connector/auth.py:678]: Can you add tests for this method?

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py]: changed

File: src/snowflake/connector/auth_keypair.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth_keypair.py]: changed

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py:693]: Added this logic, and new test case has array of accounts that go through these cases

Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py:693]: test added in test_key_pair_authentication.py

File: src/snowflake/connector/auth.py
Comment (sfc-gh-mknister) [reply src/snowflake/connector/auth.py:678]: test added in test_key_pair_authentication.py

File: test/integ/test_key_pair_authentication.py
Comment (sfc-gh-mknister) [inline test/integ/test_key_pair_authentication.py:32]: This test case generates a jwt token using the new utility function.
It then decodes the token and asserts that the fields (iss, sub, exp) contain the correct values.

I do not see a way to make a python connection with the token without changing a lot of the python connector logic, which seems too risky just for a test case. However, manual testing of the tokens using the sqlapi interface shows that they work. The test cases for making a connection with the token will be in the t_sqlapi regression test library.

**Review by sfc-gh-stakeda — APPROVED** [review]
LGTM

**Comment by sfc-gh-stakeda** [comment]
@sfc-gh-mknister , please get an approval from @sfc-gh-mkeller before merge.

File: test/integ/test_key_pair_authentication.py
Comment (sfc-gh-mkeller) [inline test/integ/test_key_pair_authentication.py:37]: You should use [parametrized tests](https://docs.pytest.org/en/stable/parametrize.html). We use them all over the code-base.

File: test/integ/test_key_pair_authentication.py
Comment (sfc-gh-mkeller) [inline test/integ/test_key_pair_authentication.py:25]: The separator embedded like that won't work on Windows
```suggestion
        current_dir, "..", "data", "rsa_keys", "rsa_key_encrypted.p8"
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

**Review by sfc-gh-stan — APPROVED** [review]
LGTM 👍

File: test/integ/test_key_pair_authentication.py
Comment (sfc-gh-mkeller) [inline test/integ/test_key_pair_authentication.py:55]: This is more of an FYI, don't do a new commit for this only.
There are these cool new features called [f-string](https://www.python.org/dev/peps/pep-0498/)
```suggestion
        f"{expected_account}.{test_user.upper()}.{public_key_fingerprint.upper()}"
```

**Review by sfc-gh-mkeller — APPROVED** [review]
🚢 🚢

**Pull request merged**
